<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2022-10-26" />

<title>Module 09: Sample QC</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization and Annotation</a>
    </li>
    <li>
      <a href="Module11X_BonusContent.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Module 09: Sample QC</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2022-10-26</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<blockquote>
<h1 id="objectives" class="unlisted unnumbered">Objectives</h1>
<ul>
<li>Generate common QC visualizations</li>
<li>Understand how to interpret QC visualizations</li>
<li>Understand when to revise the model used in the DESeq2
initialization</li>
<li>Understand the pitfalls of post-hoc analysis</li>
<li>Describe the causes and implications of batch effect or other QC
issues in an RNA-Seq experiment</li>
</ul>
</blockquote>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Prior to testing for differential expression between our comparisons
of interest, we’ll first generate plots that will assess how well our
samples match up with our expectations (based on their treatment groups)
and what we might expect to see from our differential expression
comparisons.</p>
<p><img src="images/wayfinder/slides/Slide10.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="qc-visualizations" class="section level1">
<h1>QC Visualizations</h1>
<p>We have discussed some aspects of quality control assessment at the
sequencing data level. Today we will outline sample-level and gene-level
quality control assessments to determine what we might expect for our
differential expression comparisons.</p>
<p>To do this, we will first assess the similarity of our samples by
using principal component analysis (PCA). This will allow us to
determine how well patterns in the data fits our expectations from the
experiments design and possible sources of variation.</p>
<p>Other common visualizations that we generate for our analyses include
expression heatmaps, sample correlation heatmaps, and boxplots of raw
and/or normalized counts, the code for which (due to time restrictions)
can be found as bonus content through the materials for today and in the
bonus content module.</p>
</div>
<div id="principal-component-analysis" class="section level1">
<h1>Principal Component Analysis</h1>
<p>A common and very useful plot for evaluating how well our samples
cluster by treatment groups are Principal Component Analysis (PCA)
plots. PCA is used to emphasize variation and bring out patterns in
large datasets by using dimensionality redution.</p>
<p>This image from <a
href="https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/">a
helpful step by step explaination of PCA</a> helps to illustrate the
principal component projections for two genes measured in approximately
60 mouse samples. Generally, this process is repeated and after each
gene’s contribution to a principal component or weight is determined,
the expression and weight are summed across genes for each sample to
calculate a value for each principal component.</p>
<p><img src="images/Blog_pca_6b.png" /></p>
<blockquote>
<p><strong>Note</strong>: A more detailed overview of the PCA procedure
is outlined in <a
href="https://hbctraining.github.io/DGE_workshop/lessons/principal_component_analysis.html">a
Harvard Chan Bioinformatic Core training module</a> and is based on a
more thorough description presented in a <a
href="https://www.youtube.com/watch?v=_UVHneBUBW0">StatQuest’s
video</a>. Additionally, this <a
href="https://towardsdatascience.com/principal-component-analysis-3c39fbf5cb9d">TowardsDataScience
blog post</a> goes through the math behind PCAs.</p>
</blockquote>
<div id="interpreting-pca-plots" class="section level2">
<h2>Interpreting PCA plots</h2>
<p>For most bulk RNA-seq experiments, we expect the majority of the
total variance to be explained by the first two or three principal
components.</p>
<p><img src="images/PCAplot_Fancy_rlog_ko.Tx.png"
style="width:75.0%" /></p>
<p>In this plot, which we will generate for our data later, principal
component 1 (PC1) explains ~80% of the variance in our data while
principal component 2 (PC2) explains ~12% of the variance, which fits
that expections.</p>
<blockquote>
<h1 id="question" class="unlisted unnumbered">Question</h1>
<p>How might we interpret the variance explained by each principal
component in the context of the labeled sample points?</p>
</blockquote>
<p>For more information, this <a
href="https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/">helpful
overview of PCA basics</a> walks through both the generation and
interpretation of PCA plots.</p>
</div>
<div id="evaluating-batch-effects-or-confounders"
class="section level2">
<h2>Evaluating batch effects or confounders</h2>
<p>PCA plots are also useful for evaluating the impact of factors other
than the experimental treatment or group. At times, batch effects can be
quite obvious, such as this example from the <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2
vignette</a>, where samples within each treatment group look staggered
into two subgroups.</p>
<p><img src="images/PCA1_DESeq2Vignette.png" /></p>
<p>It turns out this experiment contained samples sequenced single-end
and paired-end. If we color only by sequencing run type (paired-end
vs. single-end), we see that PC2 (29% of variance) is primarily
explained by this technical covariate.</p>
<p><img src="images/PCA2_DESeq2Vignette.png" /></p>
<p>However, the samples are clearly seperated by experimental condition
on PC1 <strong>and</strong> since we have non-confounded batches, if we
saw this pattern in our data we could incorporate the technical
covariate into our model design, such as outlined in the <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs">DESeq2
vignette</a>.</p>
<details>
<summary>
<em>Click for complex design discussion</em>
</summary>
In experiments with more complex designs, such as when there are
interesecting/multiple treatment conditions, it can be less clear what
covariants are influencing expression, such as illustrated from <a
href="http://www.molmine.com/magma/global_analysis/batch_effect.html">this
documenation for a microarray analysis tool</a>. From the PCA labeled by
experimental treatment, we see that samples from the treatment group do
not cluster together and that there is high variance across all
treatment groups. <img src="images/batch_ex1b.jpg" /> However, when
the plot is color coded by the technical batches of probe labeling, we
see that the patterns in the data are better explained by batch than the
experimental conditions. <img src="images/batch_ex1c.jpg" />
</details>
<p><br></p>
</div>
</div>
<div id="create-a-pca" class="section level1">
<h1>Create a PCA</h1>
<p>We’ve already loaded the libraries we need for this module. We have
also thought ahead in the previous module and created the
<code>outputs/figures</code> and <code>outputs/tables</code>
directories.</p>
<p>Below, we will plot the rlog normalized data and generate the PCA
projections for the top 500 using the <code>plotPCA</code> function from
DESeq2, specifying <code>condition</code> as the condition of interest,
and view the simple plot generated by the function.</p>
<pre class="r"><code>pca_plot = plotPCA(rld, intgroup = c(&#39;condition&#39;), ntop = 500)
pca_plot</code></pre>
<p><img src="fig/09-PCArlog3-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>The samples don’t appear to cluster too tightly on their
<code>condition</code>, but we do observe that they do separate in PC2.
With real data, it is often the case that data doesn’t cluster as well
as you’d expect, or that the covariate of interest is not associated
with the first (or sometimes second or third) principal component. That
doesn’t necessarily mean the experiment is a failure, but it does raise
questions such as “What is associated with PC1?” Sometimes we can’t
answer a question like this if we don’t have any sample phenotype to
color in the PCA.</p>
<p>Next, let’s save this plot as a file in our
<code>outputs/figures</code> folder. The “base R” way is to:</p>
<pre class="r"><code>pdf(file = file.path(&#39;outputs&#39;, &#39;figures&#39;, &#39;PCA_rlog_condition.pdf&#39;), width = 6, height = 6)
pca_plot
dev.off()</code></pre>
<p>Alternatively, since <code>pca_plot</code> is a <code>ggplot</code>,
we can use <code>ggsave()</code>.</p>
<pre class="r"><code>ggsave(
    filename = file.path(&#39;outputs&#39;, &#39;figures&#39;, &#39;PCA_rlog_condition.pdf&#39;),
    plot = pca_plot,
    width = 6, height = 6, units = &#39;in&#39;)</code></pre>
<blockquote>
<h1 id="exercise---customize-the-pca"
class="unlisted unnumbered">Exercise - Customize the PCA</h1>
<p>Since the <code>pca_plot</code> object is a <code>ggplot</code>–you
can see this with <code>class(pca_plot)</code>–we can use what we
learned at the end of the Computational Foundations Workshop to modify
this plot as we might see necessary. Try doing the following to the
<code>pca_plot</code>:</p>
<ol style="list-style-type: decimal">
<li>Add a title and subtitle to the plot using the <code>labs()</code>
function.</li>
<li>Change the theme of the plot with <code>theme_bw()</code>.</li>
<li>Challenge: Change the legend title to “Condition”. Hint, you can do
this with the <code>labs()</code> function too, using the corresponding
aesthetic mapping (e.g. “color”).</li>
</ol>
<p>Remember <code>ggplot2</code> adds plot components in layers, and you
can add additional layers with the <code>+</code> sign. So if the plot
already exists as an object, as <code>pca_plot</code> does in this case,
you can add to it like <code>pca_plot + ...</code></p>
</blockquote>
<details>
<summary>
Solution
</summary>
<p>Here is one possible answer:</p>
<pre class="r"><code>pca_plot +
    labs(
        title = &#39;PCA Plot&#39;,
        subtitle = &#39;Top 500 variable genes&#39;,
        color = &#39;Condition&#39;) +
    theme_bw()</code></pre>
<img src="fig/09-customize_pca_exercise-1.png" width="816" style="display: block; margin: auto;" />
</details>
<p><br></p>
<p><strong>Checkpoint</strong>: <em>If you generated and saved the
<code>pca_plot</code> PCA plot, please indicate with the green ‘check’
button. Otherwise, please use use the red ‘x’ button in your zoom
reaction panel to have this step repeated.</em></p>
<blockquote>
<h1 id="breakout-exercise-1---imagine-a-heatmap"
class="unlisted unnumbered">Breakout Exercise 1 - Imagine a heatmap</h1>
<p>Plotting the expression values across all samples for the top
variable genes in an experiment can help to visualize how samples
cluster together by their expression profiles. When combined with
phenotypic data, it can help show how samples with different treatments
behave relative to one another. Let’s work through this exercise in
small groups and generate a heatmap of the data we’ve been working
with.</p>
<p><a href="Module09a_breakout.html">Link to exercise</a></p>
</blockquote>
</div>
<div id="download-plots" class="section level1">
<h1>Download plots</h1>
<p>Rstudio server allows us to download files through the interactive
file panel on the right side. If we navigate into the plot subfolder and
select the <code>PCAplot_rlog_condition.pdf</code> or the file, we can
then click the blue gear symbol labeled <code>More</code> and select
<code>Export...</code>. We should see a prompt regarding the name of the
file and if we click <code>Download</code> the file should show up in
your local “Downloads” folder.</p>
</div>
<div id="optional-content" class="section level1">
<h1>Optional Content</h1>
<details>
<summary>
<em>Click for example code for generating a ScreePlot</em>
</summary>
<p>A screeplot is a way to visualize the variance explained by all
principal components. To generate a scree plot, the PCA results need to
be used independently of plotting, such as described by <a
href="https://statquest.org/pca-clearly-explained/">this statquest
post</a> and replicated below.</p>
<pre class="r"><code># generate PCA loadings
pca = prcomp(t(assay(rld)), scale=TRUE)

## get the scree information
pca.var = pca$sdev^2
scree = pca.var/sum(pca.var)
p = barplot((scree[1:10]*100), main=&quot;Scree Plot&quot;, xlab=&quot;Principal Component&quot;, ylab=&quot;Percent Variation&quot;)</code></pre>
<p><img src="fig/09-ScreePlot-1.png" width="816" style="display: block; margin: auto;" /></p>
<pre class="r"><code>print(p)</code></pre>
<pre><code>      [,1]
 [1,]  0.7
 [2,]  1.9
 [3,]  3.1
 [4,]  4.3
 [5,]  5.5
 [6,]  6.7
 [7,]  7.9
 [8,]  9.1
 [9,] 10.3
[10,] 11.5</code></pre>
We can see that the majority (~65%) of the variance across our samples
is explained by the first three principal components, giving us some
additional confidence regarding the quality of our data. In these scree
plot examples from BioTuring, the plot on the left fits what we would
expect for a dataset with high signal from the experimental treatment,
where the majority of the variance is explained by the first few
principal components. The plot on the right illustrates a scenario where
the variance is distributed across many components, which could be due
to low signal from the experimental treatment, complex experimental
design, or confounding factors. image: <img
src="images/proportion-of-variance-blog-horz.jpg" />
</details>
<p><br></p>
<details>
<summary>
<em>Click for code for a plot of dispersion estimates</em>
</summary>
<p>We can visualize the <strong>dispersion estimates</strong> with the
<code>plotDispEsts</code> function. This plot shows the the DESeq2
normalization results for our data, which centers on shrinking the
variance across all genes to better fit the expected spread at a given
expression level.</p>
<pre class="r"><code>plotDispEsts(dds)</code></pre>
<p><img src="fig/09-CheckDispersions-1.png" width="816" style="display: block; margin: auto;" />
Above is the raw data plotted in black, the fitted (or expected)
dispersion in red, and the normalized data with scaled variance in blue.
Since we have fairly small sample sizes for each condition, we see
shrinkage for many genes but a reasonable correlation between the
expression level and dispersions.</p>
This <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html">HBC
tutorial</a> has a more detailed overview of estimating size factors,
estimating gene dispersion, and the shrinkage procedure, as well as
examples of concerning dispersion plots that may suggest reassessing
quality of the experimental data.
</details>
<p><br></p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Discussed variance within treatment groups</li>
<li>Discussed technical artifacts, including batches</li>
<li>Learned to generate PCA plots</li>
</ul>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC QC tutorial: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html</a></li>
<li>Detailed Heatmap tutorial from Galaxy: <a
href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html"
class="uri">https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html</a></li>
<li>PCA Overview: <a
href="https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/"
class="uri">https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
</div>

<div id="rmd-source-code">---
title: "Module 09: Sample QC"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("09-")
```

> # Objectives {.unlisted .unnumbered}
> * Generate common QC visualizations
> * Understand how to interpret QC visualizations
> * Understand when to revise the model used in the DESeq2 initialization
> * Understand the pitfalls of post-hoc analysis
> * Describe the causes and implications of batch effect or other QC issues in an RNA-Seq experiment


```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
# load("rdata/RunningData.RData")
```

# Differential Expression Workflow {.unlisted .unnumbered}

Prior to testing for differential expression between our comparisons of interest, we'll first generate plots that will assess how well our samples match up with our expectations (based on their treatment groups) and what we might expect to see from our differential expression comparisons.

![](./images/wayfinder/wayfinder-SampleQCViz.png){width=75%}

---

# QC Visualizations

We have discussed some aspects of quality control assessment at the sequencing data level. Today we will outline sample-level and gene-level quality control assessments to determine what we might expect for our differential expression comparisons.

To do this, we will first assess the similarity of our samples by using principal component analysis (PCA). This will allow us to determine how well patterns in the data fits our expectations from the experiments design and possible sources of variation.

Other common visualizations that we generate for our analyses include expression heatmaps, sample correlation heatmaps, and boxplots of raw and/or normalized counts, the code for which (due to time restrictions) can be found as bonus content through the materials for today and in the bonus content module.

# Principal Component Analysis

A common and very useful plot for evaluating how well our samples cluster by treatment groups are Principal Component Analysis (PCA) plots. PCA is used to emphasize variation and bring out patterns in large datasets by using dimensionality redution.

This image from
[a helpful step by step explaination of PCA](https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/) helps to illustrate the principal component projections for two genes measured in approximately 60 mouse samples. Generally, this process is repeated and after each gene's contribution to a principal component or weight is determined, the expression and weight are summed across genes for each sample to calculate a value for each principal component.

![](./images/Blog_pca_6b.png)

>**Note**: A more detailed overview of the PCA procedure is outlined in [a Harvard Chan Bioinformatic Core training module](https://hbctraining.github.io/DGE_workshop/lessons/principal_component_analysis.html) and is based on a more thorough description presented in a [StatQuest’s video](https://www.youtube.com/watch?v=_UVHneBUBW0). Additionally, this [TowardsDataScience blog post](https://towardsdatascience.com/principal-component-analysis-3c39fbf5cb9d) goes through the math behind PCAs.

## Interpreting PCA plots

For most bulk RNA-seq experiments, we expect the majority of the total variance to be explained by the first two or three principal components.

![](images/PCAplot_Fancy_rlog_ko.Tx.png){width=75%}

In this plot, which we will generate for our data later, principal component 1 (PC1) explains ~80% of the variance in our data while principal component 2 (PC2) explains ~12% of the variance, which fits that expections.

> # Question {.unlisted .unnumbered}
>
> How might we interpret the variance explained by each principal component in the context of the labeled sample points?

For more information, this [helpful overview of PCA basics](https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/) walks through both the generation and interpretation of PCA plots.

## Evaluating batch effects or confounders

PCA plots are also useful for evaluating the impact of factors other than the experimental treatment or group. At times, batch effects can be quite obvious, such as this example from the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), where samples within each treatment group look staggered into two subgroups.

![](./images/PCA1_DESeq2Vignette.png)

It turns out this experiment contained samples sequenced single-end and paired-end. If we color only by sequencing run type (paired-end vs. single-end), we see that PC2 (29% of variance) is primarily explained by this technical covariate.

![](./images/PCA2_DESeq2Vignette.png)

However, the samples are clearly seperated by experimental condition on PC1 **and** since we have non-confounded batches, if we saw this pattern in our data we could incorporate the technical covariate into our model design, such as outlined in the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs).

<details>
    <summary>*Click for complex design discussion*</summary>
    In experiments with more complex designs, such as when there are interesecting/multiple treatment conditions, it can be less clear what covariants are influencing expression, such as illustrated from [this documenation for a microarray analysis tool](http://www.molmine.com/magma/global_analysis/batch_effect.html).
    From the PCA labeled by experimental treatment, we see that samples from the treatment group do not cluster together and that there is high variance across all treatment groups.
    ![](./images/batch_ex1b.jpg)
    However, when the plot is color coded by the technical batches of probe labeling, we see that the patterns in the data are better explained by batch than the experimental conditions.
    ![](./images/batch_ex1c.jpg)
</details>
<br>

# Create a PCA

We've already loaded the libraries we need for this module. We have also thought ahead in the previous module and created the `outputs/figures` and `outputs/tables` directories.

Below, we will plot the rlog normalized data and generate the PCA projections for the top 500 using the `plotPCA` function from DESeq2, specifying `condition` as the condition of interest, and view the simple plot generated by the function.

```{r PCArlog3}
pca_plot = plotPCA(rld, intgroup = c('condition'), ntop = 500)
pca_plot
```

The samples don't appear to cluster too tightly on their `condition`, but we do observe that they do separate in PC2. With real data, it is often the case that data doesn't cluster as well as you'd expect, or that the covariate of interest is not associated with the first (or sometimes second or third) principal component. That doesn't necessarily mean the experiment is a failure, but it does raise questions such as "What is associated with PC1?" Sometimes we can't answer a question like this if we don't have any sample phenotype to color in the PCA.

Next, let's save this plot as a file in our `outputs/figures` folder. The "base R" way is to:

```{r save_pca_base, eval = FALSE}
pdf(file = file.path('outputs', 'figures', 'PCA_rlog_condition.pdf'), width = 6, height = 6)
pca_plot
dev.off()
```

Alternatively, since `pca_plot` is a `ggplot`, we can use `ggsave()`.

```{r save_pca_ggsave, eval = FALSE}
ggsave(
    filename = file.path('outputs', 'figures', 'PCA_rlog_condition.pdf'),
    plot = pca_plot,
    width = 6, height = 6, units = 'in')
```

> # Exercise - Customize the PCA {.unlisted .unnumbered}
>
> Since the `pca_plot` object is a `ggplot`--you can see this with `class(pca_plot)`--we can use what we learned at the end of the Computational Foundations Workshop to modify this plot as we might see necessary. Try doing the following to the `pca_plot`:
>
> 1. Add a title and subtitle to the plot using the `labs()` function.
> 2. Change the theme of the plot with `theme_bw()`.
> 3. Challenge: Change the legend title to "Condition". Hint, you can do this with the `labs()` function too, using the corresponding aesthetic mapping (e.g. "color").
>
> Remember `ggplot2` adds plot components in layers, and you can add additional layers with the `+` sign. So if the plot already exists as an object, as `pca_plot` does in this case, you can add to it like `pca_plot + ...`

<details>
<summary>Solution</summary>

Here is one possible answer:

```{r customize_pca_exercise}
pca_plot +
    labs(
        title = 'PCA Plot',
        subtitle = 'Top 500 variable genes',
        color = 'Condition') +
    theme_bw()
```
</details>
<br>

**Checkpoint**: *If you generated and saved the `pca_plot` PCA plot, please indicate with the green 'check' button. Otherwise, please use  use the red 'x' button in your zoom reaction panel to have this step repeated.*

> # Breakout Exercise 1 - Imagine a heatmap {.unlisted .unnumbered}
>
> Plotting the expression values across all samples for the top variable genes in an experiment can help to visualize how samples cluster together by their expression profiles. When combined with phenotypic data, it can help show how samples with different treatments behave relative to one another. Let's work through this exercise in small groups and generate a heatmap of the data we've been working with.
>
> [Link to exercise](Module09a_breakout.html)


# Download plots

Rstudio server allows us to download files through the interactive file panel on the right side. If we navigate into the plot subfolder and select the `PCAplot_rlog_condition.pdf` or the  file, we can then click the blue gear symbol labeled `More` and select `Export...`. We should see a prompt regarding the name of the file and if we click `Download` the file should show up in your local "Downloads" folder.

# Optional Content

<details>
    <summary>*Click for example code for generating a ScreePlot*</summary>
     A screeplot is a way to visualize the variance explained by all principal components.
     To generate a scree plot, the PCA results need to be used independently of plotting, such as described by [this statquest post](https://statquest.org/pca-clearly-explained/) and replicated below.

```{r ScreePlot}
# generate PCA loadings
pca = prcomp(t(assay(rld)), scale=TRUE)

## get the scree information
pca.var = pca$sdev^2
scree = pca.var/sum(pca.var)
p = barplot((scree[1:10]*100), main="Scree Plot", xlab="Principal Component", ylab="Percent Variation")
print(p)
```

    We can see that the majority (~65%) of the variance across our samples is explained by the first three principal components, giving us some additional confidence regarding the quality of our data.
    In these scree plot examples from BioTuring, the plot on the left fits what we would expect for a dataset with high signal from the experimental treatment, where the majority of the variance is explained by the first few principal components. The plot on the right illustrates a scenario where the variance is distributed across many components, which could be due to low signal from the experimental treatment, complex experimental design, or confounding factors.
image: ![](./images/proportion-of-variance-blog-horz.jpg)
</details>
<br>

<details>
    <summary>*Click for code for a plot of dispersion estimates*</summary>
    We can visualize the **dispersion estimates** with the `plotDispEsts` function. This plot shows the the DESeq2 normalization results for our data, which centers on shrinking the variance across all genes to better fit the expected spread at a given expression level.
```{r CheckDispersions}
plotDispEsts(dds)
```
    Above is the raw data plotted in black, the fitted (or expected) dispersion in red, and the normalized data with scaled variance in blue. Since we have fairly small sample sizes for each condition, we see shrinkage for many genes but a reasonable correlation between the expression level and dispersions.

    This [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html) has a more detailed overview of estimating size factors, estimating gene dispersion, and the shrinkage procedure, as well as examples of concerning dispersion plots that may suggest reassessing quality of the experimental data.
</details>
<br>

# Summary

In this section, we:

* Discussed variance within treatment groups
* Discussed technical artifacts, including batches
* Learned to generate PCA plots

---

# Sources

* HBC QC tutorial: https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html
* Detailed Heatmap tutorial from Galaxy: https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html
* PCA Overview: https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/

```{r WriteOut.RData, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData.RData")
```

---


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module09_SampleQCViz.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
