---
title: 'Breakout Exercise 2 - Gene Annotations'
author: 'UM Bioinformatics Core'
date: '`r Sys.Date()`'
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type='text/css'>
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source('../bin/chunk-options.R')
knitr_fig_path('11a-')
```

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
```

Estimated time: **15 Minutes**

# Motivation

The data from from our differential expression test (`result_plus_vs_minus`) is an important result from our analysis.

```{r result_preview}
head(results_plus_vs_minus)
```

Looking at the first few rows, we notice that the genes are labelled with their ENSEMBL identifiers, e.g. ENSMUSG00000000056. It is quite rare that anyone will know off the top of their head which gene is associated with any given ENSEMBL identifier; we tend to think of genes by their symbols, e.g. Tp53.

# Exercise

Using the `biomaRt` package, to map ENSEMBL IDs to gene symnbols, add a column with the gene symbols to `results_plus_vs_minus`.

# Instructions

- One group member should share their screen in the breakout room. If nobody volunteers, a helper may randomly select someone.
- The group members should discuss the exercise and work together to find a solution.
- If there is time after a solution is found, allow all members to complete the exercise.

# Preliminaries

First, let's download the mapping of ENSEMBL IDs to gene symbols **together**. Then we'll split into small groups to add the gene symbol column we want. We'll start by loading the `biomaRt` library and calling the `useEnsembl()` function to select the database we'll use to extract the information we need.

```{r Pullmart, warning=FALSE}
library('biomaRt')
ensembl = useEnsembl(dataset = 'mmusculus_gene_ensembl', biomart='ensembl')
```

*Note that this process takes some time and will take up a larger amount of working memory so proceed with caution if you try to run these commands on a laptop with less than 4G of memory*

To identify possible **filters** to restrict our data, we can use the `listFilters` function. To identify the **attributes** we want to retrive, we can use the `listAttributes` function. The best approach is to use [list or search functions](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query) to help narrow down the available options.

```{r AddAnnotations2, warning=FALSE, eval=FALSE}
head(listFilters(mart = ensembl), n = 20)
head(listAttributes(ensembl), n = 30)
```

We can access additional genomic annotations using the [`bioMart` package](https://bioconductor.org/packages/release/bioc/html/biomaRt.html). To identify  we'll structure our 'query' or search of the bioMart resources to use the [ENSEMBL id](https://m.ensembl.org/info/genome/genebuild/gene_names.html) from our alignment to add the gene symbols and gene description for each gene.

```{r AddAnotation3, warning=FALSE}
id_mapping = getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
      filters = 'ensembl_gene_id',
      values = row.names(assay(dds)),
      mart = ensembl)
# will take some time to run

# Preview the result
head(id_mapping)
```

Now that we have the ENSEMBL information and a gene symbol to match to our results, we can proceed in the smaller groups. As with the previous exercise, we have broken it into small steps with hints as needed.

**Note**: For additional information regarding bioMart, please consult the [ENSEMBL bioMart vignette](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html) or the broader [Bioconductor Annotation Resources vignette ](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

# Steps

1. Look at the two data frames that are going to be needed for the exercise: `id_mapping` and `results_plus_vs_minus`.

<details>
<summary>Answer</summary>
```{r preview_tables}
head(id_mapping)
head(results_plus_vs_minus)
```
</details>
<br>

2. Think about what column of `results_plus_vs_minus` we want to match to `id_mapping`, and then what column we want to extract values from in `id_mapping`.

<details>
<summary>Answer</summary>
We want to match the `id` column of `results_plus_vs_minus` to the `ensembl_gene_id` column of `id_mapping`, and once that match is found, we want to extract the `external_gene_name` column of `id_mapping` to get the gene symbol.
</details>
<br>

3. Look at the documentation for `dplyr::left_join()` and try to merge the `id_mapping` table into the `results_plus_vs_minus` table on the columns you identified above.

<details>
<summary>Answer</summary>
```{r left_join}
results_plus_vs_minus_annotated = results_plus_vs_minus %>%
    left_join(id_mapping, by = c('id' = 'ensembl_gene_id'))
head(results_plus_vs_minus_annotated)
```
</details>
<br>

4. Optionally, how could we use some of the `tidyverse` functions we've encountered previously to rename the `external_gene_name` column to `symbol` and to move it into the second column position? Hint: Because of the order of the packages we may have loaded, consider using `dplyr::rename()` and `dplyr::select()` instead of just the `select()` function. We can discuss why this is the case together.

<details>
<summary>Answer</summary>
```{r rename_rearrange}
results_plus_vs_minus_annotated = results_plus_vs_minus_annotated %>%
    dplyr::rename('symbol' = 'external_gene_name') %>%
    dplyr::select(id, symbol, everything())
results_plus_vs_minus_annotated
```
</details>
<br>


```{r WriteOut.RData, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = 'rdata/RunningData.RData')
```
